feat: Support Parallel Segment Fetching

tyrelltle opened the pull request:
closes https://github.com/shaka-project/shaka-player/issues/4658.

This solution is inspired by abandoned PR https://github.com/shaka-project/shaka-player/pull/2809, which implements segment prefetching ahead of current play head. 

![image](https://user-images.githubusercontent.com/3315733/205465795-75c605d2-c2e3-4d03-90f5-46a72a7189d2.png)

avelad commented:
Thanks for your patience, I'll try to review it throughout this week.

avelad commented on externs/shaka/player.js:
```diff
@@ -1077,6 +1078,12 @@ shaka.extern.ManifestConfiguration;
+ *   ahead of playhead in parallel.
+ *   If <code>null</code> or <code><1</code>, the segments will be fetched
+ *   sequentially.
+ *   Defaults to <code>null</code>.
```
Defaults to <code>0</code>.

tyrelltle replied:
updated

avelad commented on externs/shaka/player.js:
```diff
@@ -1077,6 +1078,12 @@ shaka.extern.ManifestConfiguration;
+ * @property {boolean} segmentPrefetchLimit
+ *   The maximum number of segments for each active stream to be prefetched
+ *   ahead of playhead in parallel.
+ *   If <code>null</code> or <code><1</code>, the segments will be fetched
```
null is not a valid value

tyrelltle replied:
changed to 0

avelad commented on lib/media/streaming_engine.js:
```diff
@@ -170,6 +171,23 @@ shaka.media.StreamingEngine = class {
+        state.segmentPrefetch.resetLimit(config.segmentPrefetchLimit);
+        if (!(config.segmentPrefetchLimit > 0)) {
+          /**
+           * ResetLimit is still needed in this case,
```
Please use a // comment

tyrelltle replied:
updated

avelad commented on lib/media/streaming_engine.js:
```diff
@@ -1091,12 +1133,15 @@ shaka.media.StreamingEngine = class {
     return null;
   }
 
-
```
Please revert this change

tyrelltle replied:
updated

avelad commented on lib/media/streaming_engine.js:
```diff
@@ -2000,6 +2045,35 @@ shaka.media.StreamingEngine = class {
+   *
+   * @return {!shaka.net.NetworkingEngine.PendingRequest}
+   * @private
+   * @suppress {strictMissingProperties}
```
Why?

tyrelltle replied:
It was originally copied from fetch_() as this method was extracted from that.
I just tried remove it but got following err. Added instanceof check and working now.


```
[JSC_STRICT_INEXISTENT_UNION_PROPERTY] Property startTime not defined on all member types of (shaka.media.InitSegmentReference|shaka.media.SegmentReference)
  2091|           duration: reference.endTime - reference.startTime,
                                                          ^^^^^^^^^
```

joeyparrish commented:
I skimmed this PR, and it looks interesting and well-constructed.  You've designed something that appears to address most of my concerns, it defaults to 0 (good for low-memory devices), and the prefetcher cleans up after itself so it doesn't hold onto an unbounded number of segments.  This could also be useful in a future implementation of #880 (preload API).

I haven't read it in detail, and I have some deadlines to get AV sync fixes done this week before we go on holidays at Google.  But I will come back to this in January for a detailed review, and I expect this is something that will ultimately ship in Shaka Player.

Thanks for your contribution!

tyrelltle commented:
Thanks @joeyparrish !
Assuming this change goes into main branch in January, when a new version that has the change will be released ?
Asking so we can plan accordingly in case the change is not expected to be released soon.
Thanks !

joeyparrish commented:
We can make a feature release at any time.  We don't have a fixed release cycle.

avelad commented:
@tyrelltle can you rebase and resolve the conflicts? Thanks!

tyrelltle commented:
@avelad yes I just rebased now

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+   * @param {function(
+   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+   *  shaka.extern.Stream
+   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
```
I would define the `fetchDispatcher` function with a `@typedef`, so it doesn't need to be written out each time like this.
See, for reference, `shaka.net.NetworkingEngine.OnHeadersReceived` inside `lib/net/networking_engine.js`.

tyrelltle replied:
defined type

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+   */
+  constructor(prefetchLimit, stream, fetchDispatcher) {
+    /**
+     * @private {number}
```
This block could be single-line.

tyrelltle replied:
updated

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+
+    /**
+     * @private {!Map.<shaka.media.SegmentReference,
+                 !shaka.net.NetworkingEngine.PendingRequest>}
```
This second line of the comment here should have an asterisk at the start too.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+   * @public
+   */
+  prefetchSegments(startReference) {
+    if (!(this.prefetchLimit_ > 0)) {
```
`prefetchLimit_` is non-nullable, so couldn't this just be:
```js
if (this.prefetchLimit_ <= 0) {
```

tyrelltle replied:
replaced with assert

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+    if (!(this.prefetchLimit_ > 0)) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
This could be compacted into one line:
```js
const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(this.stream_);
```

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+        this.stream_,
+    );
+    if (!this.stream_.segmentIndex) {
+      shaka.log.info(
```
This could be compacted into one line too.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+   */
+  getPrefetchedSegment(reference) {
+    let op = null;
+    if (
```
I assume that returning null here if the input is not a SegmentReference is to leave open the possibility of fetching init segments at a later date?
If so, it should probably be a separate return clause, with a comment explaining that it's a placeholder.
That would also remove the need for this multi-line if-statement.

tyrelltle replied:
actually it was a mistake to leave InitSegmentReferece as an input type, so i changed the input type to only be SegmentReference

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+      return null;
+    }
+
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
This could be compacted into one line.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+    );
+
+    if (this.segmentPrefetchMap_.has(reference)) {
+      op = this.segmentPrefetchMap_.get(reference);
```
I don't think you really need to define the `op` variable at the top of this function, since it's only set here anyway.
You could just declare it as a local variable in this clause, and the other clause could just return null.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+        this.segmentPrefetchMap_.set(reference, op);
+      }
+      this.prefetchPosTime_ = reference.startTime;
+      reference =iterator.next().value;
```
There should be a space after the `=`.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+    if (this.segmentPrefetchMap_.size === 0) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
```
This could be compacted into one line.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+    }
+  }
+
+  /** @private */
```
This JSDoc should list the input parameter, and the return value.

theodab commented on lib/media/segment_prefetch.js:
```diff
@@ -0,0 +1,173 @@
+   */
+  resetLimit(prefetchLimit) {
+    if (prefetchLimit !== this.prefetchLimit_) {
+      this.clearAll();
```
Design-wise, I'm not sure it makes sense to always clear all of the segments away in this case.
For example, if you are going from a pre-fetch limit of 4 to a pre-fetch limit of 6, couldn't you just keep the stuff you have already pre-fetched?

It seems like the most efficient thing to do would be to store an array that keeps track of the order of the pre-fetched segments, and then pop and clear segments from the end of it until you fit into the new limit. That way nothing happens if the limit has been increased, and you clear away as few pre-fetched segments as possible in the case of the limit being decreased.

tyrelltle replied:
updated. actually it turns out that Map.keys reserves the insertion order.
so instead of creating a new ordered array, i just grab `segmentPrefetchMap_.keys` and iterate from end to pop + clear the segments.
not sure Array.from(iterator) will actually do any iteration though, but it might have minimal performance impact since resetLimit could be called very rarely.
open to other suggestions :)

tyrelltle replied:
pushed a new commit to simplify resetLimit, by using while + arr.pop() instead of for loop from end of arr

theodab commented on lib/media/streaming_engine.js:
```diff
@@ -818,6 +838,7 @@ shaka.media.StreamingEngine = class {
       stream,
       type: stream.type,
       segmentIterator: null,
+      segmentPrefetch: this.createSegmentPrefetch_(stream),
```
We probably should only make prefetchers for video and audio streams, not text or image.

tyrelltle replied:
updated createSegmentPrefetch_ to return null if stream.type is not VIDEO nor AUDIO

theodab commented on lib/media/streaming_engine.js:
```diff
@@ -170,6 +171,21 @@ shaka.media.StreamingEngine = class {
+        if (!(config.segmentPrefetchLimit > 0)) {
+          // ResetLimit is still needed in this case,
+          // to abort existing prefetch operations.
+          state.segmentPrefetch = null;
```
If we clear the `segmentPrefetch` object away if the limit is reduced to 0, we might not need all of those if-statements that check to see if the limit is >0 inside the class itself.
Instead, you should turn those into asserts.

tyrelltle replied:
changed the "if prefetchLimit < 0" statements that are inside SegmentPrefetch, to goog.asserts.assert

theodab commented on lib/media/streaming_engine.js:
```diff
@@ -2027,28 +2099,25 @@ shaka.media.StreamingEngine = class {
-
-    const stream = mediaState.stream;
+    let duration = 0;
+    if ((reference instanceof shaka.media.SegmentReference)) {
```
You don't need the second set of parentheses here.

theodab commented on lib/media/streaming_engine.js:
```diff
@@ -2336,6 +2414,8 @@ shaka.media.StreamingEngine.PlayerInterface;
  * @property {shaka.net.NetworkingEngine.PendingRequest} operation
  *   Operation with the number of bytes to be downloaded.
+ * @property {shaka.media.SegmentPrefetch} segmentPrefetch
+ *   segmentPrefetch.
```
You could use an actual description here.
Perhaps something like...

> A prefetch object for managing prefetching. Null if unneeded (if prefetching is disabled, etc).

tyrelltle replied:
updated

theodab commented on lib/media/streaming_engine.js:
```diff
@@ -2336,6 +2414,8 @@ shaka.media.StreamingEngine.PlayerInterface;
  *   updating.
  * @property {shaka.net.NetworkingEngine.PendingRequest} operation
  *   Operation with the number of bytes to be downloaded.
+ * @property {shaka.media.SegmentPrefetch} segmentPrefetch
```
This should probably be `@property {?shaka.media.SegmentPrefetch}`, to make it clearer that this is expected to sometimes be null.

tyrelltle replied:
updated

theodab commented on test/test/util/fake_segment_prefetch.js:
```diff
@@ -0,0 +1,81 @@
+    /** @private {shaka.extern.Stream} */
+    this.streamObj_ = stream;
+
+    /** @private {
```
The `/**` and `*/` should be on separate lines. Also, this could be compacted some.

```js
   /**
    * @private {!Object.<string, shaka.test.FakeMediaSourceEngine.SegmentData>}
    */
```

tyrelltle commented:
@theodab thanks for the comments !
Agree all of them and will make sure address them early next week

tyrelltle commented on demo/locales/source.json:
```diff
@@ -1042,5 +1042,9 @@
     "message": "Filters for assets that have [JARGON:XLINK] tags in their manifests, so that they can be broken into multiple files."
+  },
+  "DEMO_SEGMENT_PREFETCH_LIMIT": {
+    "description": "Segment Prefetch Limit configuration used by StreamingEngine.",
```
updated to not mention StreamingEngine

tyrelltle commented:
> @theodab thanks for the comments ! Agree all of them and will make sure address them early next week

addressed some of the comments, will finish other ones by tomorrow

theodab reviewed (changes requested):
Thanks for the work, by the way.

tyrelltle commented on lib/media/segment_prefetch.js:
```diff
@@ -20,15 +21,10 @@ shaka.media.SegmentPrefetch = class {
-   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
-   *  shaka.extern.Stream
-   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
+   * @param {shaka.media.SegmentPrefetch.fetchDispatcher} fetchDispatcher
```
sorry missed that, updated

tyrelltle commented:
> Thanks for the work, by the way.

Thanks for the comments I updated all of them. Let me know if any other suggestions :)
